File: data/src/main/java/com/github/mrmitew/bodylog/data/repository/ProfileRepository.java
package com.github.mrmitew.bodylog.data.repository;

import com.github.mrmitew.bodylog.domain.repository.Repository;
import com.github.mrmitew.bodylog.domain.repository.entity.Profile;
import com.jakewharton.rxrelay2.BehaviorRelay;

import org.jetbrains.annotations.NotNull;

import java.util.concurrent.TimeUnit;

import io.reactivex.Completable;
import io.reactivex.Observable;

public class ProfileRepository implements Repository {
    private final BehaviorRelay<Profile> mProfileBehaviorRelay;

    private static Profile sCachedProfile = Profile.builder()
            .name("John Doe")
            .description("With hard work you can achieve your goals and you can become successful!")
            .weight(70)
            .bodyFat(8)
            .backSize(120)
            .chestSize(100)
            .armsSize(40)
            .waistSize(74.5f)
            .build();

    public ProfileRepository() {
        mProfileBehaviorRelay = BehaviorRelay.create();
    }

    @Override
    @NotNull
    public Observable<Profile> getProfile() {
        return Observable.just(sCachedProfile)
                // Simulate a long process
                .delay(1500, TimeUnit.MILLISECONDS);
    }

    @Override
    @NotNull
    public Observable<Profile> getProfileRefreshing() {
        return mProfileBehaviorRelay.startWith(sCachedProfile)
                // Simulate a long process
                .delay(1500, TimeUnit.MILLISECONDS);
    }

    @Override
    @NotNull
    public Completable setProfile(@NotNull final Profile profile) {
        return Completable.fromAction(() -> sCachedProfile = profile)
                .doOnComplete(() -> mProfileBehaviorRelay.accept(profile));
    }
}

File: data/src/main/java/com/github/mrmitew/bodylog/data/repository/ProfileRepository.kt
package com.github.mrmitew.bodylog.data.repository

import com.github.mrmitew.bodylog.domain.repository.Repository
import com.github.mrmitew.bodylog.domain.repository.entity.Profile
import com.jakewharton.rxrelay2.BehaviorRelay
import io.reactivex.Completable
import io.reactivex.Observable
import java.util.concurrent.TimeUnit

class ProfileRepository : Repository {
    private val mProfileBehaviorRelay: BehaviorRelay<Profile> = BehaviorRelay.create()
    private var sCachedProfile = Profile(
            name = "John Doe",
            description = "With hard work you can achieve your goals and you can become successful!",
            weight = 70f,
            bodyFatPercentage = 8f,
            backSize = 120f,
            chestSize = 100f,
            armsSize = 40f,
            waistSize = 74.5f)

    override fun getProfile(): Observable<Profile> {
        return Observable.just<Profile>(sCachedProfile)
                // Simulate a long process
                .delay(1500, TimeUnit.MILLISECONDS)
    }

    override fun getProfileRefreshing(): Observable<Profile> {
        return mProfileBehaviorRelay.startWith(sCachedProfile)
                // Simulate a long process
                .delay(1500, TimeUnit.MILLISECONDS)
    }

    override fun setProfile(profile: Profile): Completable {
        return Completable.fromAction { sCachedProfile = profile }
                .doOnComplete { mProfileBehaviorRelay.accept(profile) }
    }
}
